File: CostUI/src/app/services/cost.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams} from '@angular/common/http';
import { CostItem } from '../models/cost-item';
import { ChartData } from '../models/chart-data';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class CostService {
  private apiUrl = 'http://localhost:5000/api/costs';

  constructor(private http: HttpClient) {}

  getCosts(
    start: string,
    end: string,
    incomexpense: string,
    item: string,
    paymethod: string
  ): Observable<CostItem[]> {
    let params = new HttpParams()
      .set('start', start)
      .set('end', end);

    if (incomexpense !== '全部') {
      params = params.set('incomexpense', incomexpense);
    }

    if (item !== '全部') {
      params = params.set('item', item);
    }

    if (paymethod !== '全部') {
      params = params.set('paymethod', paymethod);
    }

    return this.http.get<CostItem[]>(this.apiUrl, { params });
  }

  addCost(cost: CostItem): Observable<void> {
    return this.http.post<void>(this.apiUrl, cost);
  }


}

File: CostApi_EF/Services/CostService.cs
using CostApi_EF.Data;
using CostApi_EF.Models;
using Microsoft.EntityFrameworkCore;

namespace CostApi_EF.Services;

// 原本的 CsvService 讀寫檔案邏輯，改成用 EF Core 操作資料庫。
// 介面（ICostService）完全沒變，所以 CostsController 幾乎不用改。
public class CostService : ICostService
{
    private readonly CostDbContext _db;

    public CostService(CostDbContext db)
    {
        _db = db;
    }

    public void AddCost(CostItem item)
    {
        item.Id = 0; // Id 由資料庫自動編號，強制歸零避免被誤傳成更新既有資料
        _db.Costs.Add(item);
        _db.SaveChanges();
    }

    public List<CostItem> GetCosts(
        DateOnly start,
        DateOnly end,
        string incomexpense,
        string item,
        string paymethod)
    {
        var query = _db.Costs.Where(c => c.Datetime >= start && c.Datetime <= end);

        if (incomexpense != "全部")
        {
            query = query.Where(c => c.Incomexpense == incomexpense);
        }

        if (item != "全部")
        {
            query = query.Where(c => c.Item == item);
        }

        if (paymethod != "全部")
        {
            query = query.Where(c => c.Paymethod == paymethod);
        }

        return query.OrderByDescending(c => c.Datetime).ToList();
    }
}
